### PR TITLE
Feature: Semantic Zoom

### PR DESCRIPTION
**Pull request by elliotgordonrowe**
Takes in an array of Objects each containing zoomLevel and options properties. The options property is an object where all the keys are options and the values are desired settings below that zoom level and above the lower one. Also adds a zoomChange callback that triggers every time the window scale changes.

### CODE REVIEW

File: src/Builder.jsx [hunk 1]
```diff
@@ -90,6 +91,22 @@ class Builder {
       full_screen_button: false,
       ignore_bootstrap: false,
       disabled_buttons: null,
+      semanticZoom: [
+        {
+          zoomLevel: 0.3,
```
Comment (zakandrewking) [inline src/Builder.jsx]: let's stick with snake case (`zoom_level`) to match the rest of the API

File: src/Builder.jsx [hunk 2]
```diff
@@ -497,7 +543,16 @@ class Builder {
...
           settings={this.settings}
-          saveMap={() => this.map.save()}
+          saveMap={() => {
+            if (this.semanticOptions) {
+              Object.entries(this.semanticOptions).map(([key, value]) => {
+                this.settings.set_conditional(key, value)
+              })
+              this._update_data()
+            }
+            console.log(this.options.hide_all_labels)
```
Comment (zakandrewking) [inline src/Builder.jsx]: drop this log

File: src/Builder.jsx [hunk 3]
```diff
@@ -497,7 +543,16 @@ class Builder {
           {...this.map}
           mode={mode}
           settings={this.settings}
-          saveMap={() => this.map.save()}
+          saveMap={() => {
+            if (this.semanticOptions) {
```
Comment (zakandrewking) [inline src/Builder.jsx]: can you comment on the purpose of this logic?

Comment (elliotgordonrowe) [reply src/Builder.jsx]: Added a comment explaining why I did it but basically if semanticZoom is active it reverts the options to their original loaded values before saving

**Pull request merged**